File: project01-unidb-ahk1384/templates/python/main.py
from abc import ABC, abstractmethod
from typing import List, Optional

# ==========================================
# Domain Models
# ==========================================
class Student:
    def __init__(self, id: int, name: str, gpa: float):
        self.id = id
        self.name = name
        self.gpa = gpa

    def __repr__(self):
        return f"{{id: {self.id}, name: '{self.name}', gpa: {self.gpa}}}"

# ==========================================
# Layer 3: Storage Engine (Interface)
# ==========================================
class Collection(ABC):
    @abstractmethod
    def insert_one(self, student: Student):
        pass

    @abstractmethod
    def delete_one(self, id: int):
        pass

    @abstractmethod
    def find_by_id(self, id: int) -> Optional[Student]:
        pass

    @abstractmethod
    def find_all(self) -> List[Student]:
        pass

class ArrayCollection(Collection):
    def __init__(self):
        self.data = []  # Python list is a dynamic array

    def insert_one(self, student: Student):
        # TODO: Implement
        pass

    def delete_one(self, id: int):
        # TODO: Implement
        pass

    def find_by_id(self, id: int) -> Optional[Student]:
        # TODO: Implement
        return None

    def find_all(self) -> List[Student]:
        return []

class LinkedListCollection(Collection):
    class Node:
        def __init__(self, data):
            self.data = data
            self.next = None
            self.prev = None

    def __init__(self):
        self.head = None
        self.tail = None

    def insert_one(self, student: Student):
        # TODO: Implement
        pass

    def delete_one(self, id: int):
        # TODO: Implement
        pass

    def find_by_id(self, id: int) -> Optional[Student]:
        # TODO: Implement
        return None

    def find_all(self) -> List[Student]:
        return []

# ==========================================
# Layer 2: Execution Engine
# ==========================================
class ExecutionEngine:
    def __init__(self):
        self.current_collection = ArrayCollection()
        # TODO: self.transaction_stack = []
        # TODO: self.batch_queue = []

    def execute_command(self, command_type: str, args: List[str]):
        if command_type == "insertOne":
            print("Executing insertOne...")
            # TODO: Parse args
        elif command_type == "findByID":
            pass
        # TODO: Handle others

# ==========================================
# Layer 1: Query Parser
# ==========================================
class QueryParser:
    @staticmethod
    def parse_and_execute(input_str: str, engine: ExecutionEngine):
        tokens = input_str.split()
        if not tokens:
            return

        # TODO: Improve parsing logic
        # Example: db.students.insertOne({...}) -> extract 'insertOne' and json
        cmd = tokens[0]
        engine.execute_command(cmd, tokens[1:])

# ==========================================
# Main Loop
# ==========================================
if __name__ == "__main__":
    engine = ExecutionEngine()
    print("UniDB Shell (Python)\nType 'exit' to quit.")

    while True:
        try:
            user_input = input("unidb> ")
            if user_input == "exit":
                break
            QueryParser.parse_and_execute(user_input, engine)
        except KeyboardInterrupt:
            break

File: project01-unidb-ahk1384/templates/java/src/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

// ==========================================
// Domain Models
// ==========================================
class Student {
    public int id;
    public String name;
    public double gpa;

    public Student(int id, String name, double gpa) {
        this.id = id;
        this.name = name;
        this.gpa = gpa;
    }
}

// ==========================================
// Layer 3: Storage Engine (Interface)
// ==========================================
interface Collection {
    void insertOne(Student student);
    void deleteOne(int id);
    Student findByID(int id);
    List<Student> findAll();
}

class ArrayCollection implements Collection {
    // TODO: Use ArrayList or raw array
    public void insertOne(Student student) {
        // TODO: Implement
    }
    public void deleteOne(int id) {
        // TODO: Implement
    }
    public Student findByID(int id) {
        // TODO: Implement
        return null;
    }
    public List<Student> findAll() {
        return new ArrayList<>();
    }
}

class LinkedListCollection implements Collection {
    private class Node {
        Student data;
        Node next;
        Node prev;
    }
    // TODO: Manage head/tail

    public void insertOne(Student student) {
        // TODO: Implement
    }
    public void deleteOne(int id) {
        // TODO: Implement
    }
    public Student findByID(int id) {
        // TODO: Implement
        return null;
    }
    public List<Student> findAll() {
        return new ArrayList<>();
    }
}

// ==========================================
// Layer 2: Execution Engine
// ==========================================
class ExecutionEngine {
    private Collection currentCollection;
    // TODO: Add Stack<Command> for Transactions
    // TODO: Add Queue<Command> for Batch

    public ExecutionEngine() {
        this.currentCollection = new ArrayCollection();
    }

    public void executeCommand(String commandType, String[] args) {
        if (commandType.equals("insertOne")) {
            // TODO: Parse JSON-like args
            System.out.println("Executing insertOne...");
        }
        // TODO: Handle others
    }
}

// ==========================================
// Layer 1: Query Parser
// ==========================================
class QueryParser {
    public static void parseAndExecute(String input, ExecutionEngine engine) {
        // Simple tokenizer
        String[] tokens = input.split(" ");
        if (tokens.length == 0) return;

        // TODO: Improve parsing logic to handle 'db.students.insertOne({...})'
        // Current logic is just a placeholder
        String cmd = tokens[0];
        // Pass args...
        engine.executeCommand(cmd, tokens);
    }
}

// ==========================================
// Main Loop (Client Simulation)
// ==========================================
public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        ExecutionEngine engine = new ExecutionEngine();

        System.out.println("UniDB Shell (Java)\nType 'exit' to quit.");

        while (true) {
            System.out.print("unidb> ");
            String input = scanner.nextLine();
            if (input.equals("exit")) break;

            QueryParser.parseAndExecute(input, engine);
        }
    }
}

